### ISSUE TITLE
Does introspect cause memory leak?

### ISSUE THREAD

**Issue by okmyself** [siteline/SwiftUI-Introspect#161]
@crayment When I use introspect in a struct containing a viewModel, the deinit function in the viewModel never executes.
Version: Master branch on github (2022-8-25), I downloaded this to fix #140, but found this issue instead.

**Comment by okmyself** [comment]
After modifying to the following code, the memory leak problem seems to be solved.

**Comment by crayment** [comment]
Thanks for reporting that! If you could open a pull request with that change I would put my support behind it and pull it into my fork. This is not my repo though so I would not be able to merge it here.

**Comment by okmyself** [comment]
> 

Thank you. I will try it.

**Comment by okmyself** [comment]
> Thanks for reporting that! If you could open a pull request with that change I would put my support behind it and pull it into my fork. This is not my repo though so I would not be able to merge it here.

I have completed my pull requests. Is there anything else I need to do?

**Comment by okmyself** [comment]
> Thanks for reporting that! If you could open a pull request with that change I would put my support behind it and pull it into my fork. This is not my repo though so I would not be able to merge it here.

@crayment I have completed my pull requests. Is there anything else I need to do?

**Comment by crayment** [comment]
Thanks or doing that! That should be all - but in my experience in might be a while before it gets reviewed/merged unfortunately.

**Comment by plaps153** [comment]
This really works. It solved my problem and should be merged!

**Comment by davdroman** [comment]
@okmyself @crayment is this still a problem? I'm happy to revisit #163 and get it merged if this issue is still ongoing.

**Comment by okmyself** [comment]
I have fixed it myself. Use the code below:


UIKitIntrospectView.swift

#if canImport(UIKit)

import UIKit

import SwiftUI




/// Introspection UIView that is inserted alongside the target view.

@available(iOS 13.0, *)

public class IntrospectionUIView: UIView {

&nbsp;&nbsp; &nbsp;

&nbsp; &nbsp; var moveToWindowHandler: (() -&gt; Void)?

&nbsp;&nbsp; &nbsp;

&nbsp; &nbsp; required init() {

&nbsp; &nbsp; &nbsp; &nbsp; super.init(frame: .zero)

&nbsp; &nbsp; &nbsp; &nbsp; isHidden = true

&nbsp; &nbsp; &nbsp; &nbsp; isUserInteractionEnabled = false

&nbsp; &nbsp; }

&nbsp;&nbsp; &nbsp;

&nbsp; &nbsp; @available(*, unavailable)

&nbsp; &nbsp; required init?(coder: NSCoder) {

&nbsp; &nbsp; &nbsp; &nbsp; fatalError("init(coder:) has not been implemented")

&nbsp; &nbsp; }

&nbsp;&nbsp; &nbsp;

&nbsp; &nbsp; override public func didMoveToWindow() {

&nbsp; &nbsp; &nbsp; &nbsp; super.didMoveToWindow()

&nbsp; &nbsp; &nbsp; &nbsp; moveToWindowHandler?()

&nbsp; &nbsp; }

}




/// Introspection View that is injected into the UIKit hierarchy alongside the target view.

/// After `updateUIView` is called, it calls `selector` to find the target view, then `customize` when the target view is found.

@available(iOS 13.0, tvOS 13.0, macOS 10.15.0, *)

public struct UIKitIntrospectionView<TargetViewType: UIView&gt;: UIViewRepresentable {

&nbsp;&nbsp; &nbsp;

&nbsp; &nbsp; /// Method that introspects the view hierarchy to find the target view.

&nbsp; &nbsp; /// First argument is the introspection view itself, which is contained in a view host alongside the target view.

&nbsp; &nbsp; let selector: (IntrospectionUIView) -&gt; TargetViewType?

&nbsp;&nbsp; &nbsp;

&nbsp; &nbsp; /// User-provided customization method for the target view.

&nbsp; &nbsp; let customize: (TargetViewType) -&gt; Void

&nbsp;&nbsp; &nbsp;

&nbsp; &nbsp; public init(

&nbsp; &nbsp; &nbsp; &nbsp; selector: @escaping (IntrospectionUIView) -&gt; TargetViewType?,

&nbsp; &nbsp; &nbsp; &nbsp; customize: @escaping (TargetViewType) -&gt; Void

&nbsp; &nbsp; ) {

&nbsp; &nbsp; &nbsp; &nbsp; self.selector = selector

&nbsp; &nbsp; &nbsp; &nbsp; self.customize = customize

&nbsp; &nbsp; }

&nbsp;&nbsp; &nbsp;

&nbsp; &nbsp; public func makeUIView(context: UIViewRepresentableContext<UIKitIntrospectionView&gt;) -&gt; IntrospectionUIView {

&nbsp; &nbsp; &nbsp; &nbsp; let view = IntrospectionUIView()

&nbsp; &nbsp; &nbsp; &nbsp; view.accessibilityLabel = "IntrospectionUIView<\(TargetViewType.self)&gt;"

&nbsp; &nbsp; &nbsp; &nbsp; view.moveToWindowHandler = {

&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; DispatchQueue.main.async {

&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; guard let targetView = selector(view) else {

&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; return

&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; }

&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; customize(targetView)

&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; }

&nbsp; &nbsp; &nbsp; &nbsp; }

&nbsp; &nbsp; &nbsp; &nbsp; return view

&nbsp; &nbsp; }




&nbsp; &nbsp; /// When `updateUiView` is called after creating the Introspection view, it is not yet in the UIKit hierarchy.

&nbsp; &nbsp; /// At this point, `introspectionView.superview.superview` is nil and we can't access the target UIKit view.

&nbsp; &nbsp; /// To workaround this, we wait until the runloop is done inserting the introspection view in the hierarchy, then run the selector.

&nbsp; &nbsp; /// Finding the target view fails silently if the selector yield no result. This happens when `updateUIView`

&nbsp; &nbsp; /// gets called when the introspection view gets removed from the hierarchy.

&nbsp; &nbsp; public func updateUIView(

&nbsp; &nbsp; &nbsp; &nbsp; _ uiView: IntrospectionUIView,

&nbsp; &nbsp; &nbsp; &nbsp; context: UIViewRepresentableContext<UIKitIntrospectionView&gt;

&nbsp; &nbsp; ) {

&nbsp; &nbsp; &nbsp; &nbsp; uiView.moveToWindowHandler?()

&nbsp; &nbsp; }

&nbsp;&nbsp; &nbsp;

&nbsp; &nbsp; /// Fix memory leak issue

&nbsp; &nbsp; public static func dismantleUIView(_ uiView: IntrospectionUIView, coordinator: ()) {

&nbsp; &nbsp; &nbsp; &nbsp; uiView.moveToWindowHandler = nil

&nbsp; &nbsp; }

}

#endif




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "siteline/SwiftUI-Introspect"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年2月11日(星期六) 上午9:16
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [siteline/SwiftUI-Introspect] Does introspect cause memory leak? (Issue #161)





 
@okmyself @crayment is this still a problem? I'm happy to revisit #163 and get it merged if this issue is still ongoing.

**Comment by okmyself** [comment]
It is still a problem, and I have found another problem yet. In some cases UIView does not call moveToWindowHandler automatically, so the moveToWindowHandler call in updateUIView must be preserved, please see another email for detailed code.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "siteline/SwiftUI-Introspect"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年2月11日(星期六) 上午9:16
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [siteline/SwiftUI-Introspect] Does introspect cause memory leak? (Issue #161)





 
@okmyself @crayment is this still a problem? I'm happy to revisit #163 and get it merged if this issue is still ongoing.

**Comment by davdroman** [comment]
@okmyself please comment via the GitHub web interface as GitHub does a [terrible job](https://github.com/siteline/SwiftUI-Introspect/issues/161#issuecomment-1426856587) at rendering email responses.

**Comment by okmyself** [comment]
> @okmyself please comment via the GitHub web interface as GitHub does a [terrible job](https://github.com/siteline/SwiftUI-Introspect/issues/161#issuecomment-1426856587) at rendering email responses.

This issue was introduced by bug fix #153. The version displayed on Github still has issues, but the current official version does not have this problem. I have also discovered another issue. In some cases, UIView does not call moveToWindowHandle, so the moveToWindowHandler call in updateUIView must be preserved.






UIKitIntrospectionView.swift
```
#if canImport(UIKit)
import UIKit
import SwiftUI

/// Introspection UIView that is inserted alongside the target view.
@available(iOS 13.0, *)
public class IntrospectionUIView: UIView {
    
    var moveToWindowHandler: (() -> Void)?
    
    required init() {
        super.init(frame: .zero)
        isHidden = true
        isUserInteractionEnabled = false
    }
    
    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override public func didMoveToWindow() {
        super.didMoveToWindow()
        moveToWindowHandler?()
    }
}

/// Introspection View that is injected into the UIKit hierarchy alongside the target view.
/// After `updateUIView` is called, it calls `selector` to find the target view, then `customize` when the target view is found.
@available(iOS 13.0, tvOS 13.0, macOS 10.15.0, *)
public struct UIKitIntrospectionView<TargetViewType: UIView>: UIViewRepresentable {
    
    /// Method that introspects the view hierarchy to find the target view.
    /// First argument is the introspection view itself, which is contained in a view host alongside the target view.
    let selector: (IntrospectionUIView) -> TargetViewType?
    
    /// User-provided customization method for the target view.
    let customize: (TargetViewType) -> Void
    
    public init(
        selector: @escaping (IntrospectionUIView) -> TargetViewType?,
        customize: @escaping (TargetViewType) -> Void
    ) {
        self.selector = selector
        self.customize = customize
    }
    
    public func makeUIView(context: UIViewRepresentableContext<UIKitIntrospectionView>) -> IntrospectionUIView {
        let view = IntrospectionUIView()
        view.accessibilityLabel = "IntrospectionUIView<\(TargetViewType.self)>"
        view.moveToWindowHandler = {
            DispatchQueue.main.async {
                guard let targetView = selector(view) else {
                    return
                }
                customize(targetView)
            }
        }
        return view
    }

    /// When `updateUiView` is called after creating the Introspection view, it is not yet in the UIKit hierarchy.
    /// At this point, `introspectionView.superview.superview` is nil and we can't access the target UIKit view.
    /// To workaround this, we wait until the runloop is done inserting the introspection view in the hierarchy, then run the selector.
    /// Finding the target view fails silently if the selector yield no result. This happens when `updateUIView`
    /// gets called when the introspection view gets removed from the hierarchy.
    public func updateUIView(
        _ uiView: IntrospectionUIView,
        context: UIViewRepresentableContext<UIKitIntrospectionView>
    ) {
        uiView.moveToWindowHandler?()
    }
    
    /// Fix memory leak issue
    public static func dismantleUIView(_ uiView: IntrospectionUIView, coordinator: ()) {
        uiView.moveToWindowHandler = nil
    }
}
#endif
```

**Comment by davdroman** [comment]
@okmyself okay I understand. Would you be willing to put forth a PR? I'll merge it right away. Otherwise I can do it when I get some spare time, but I'd appreciate if you could do it as I have lots of other issues to go through on this repo. Regardless: thank you for your responses!
